source/simulator/stokes_matrix_free.cc

```diff
@@ -1973,12 +1998,21 @@ namespace aspect
     solver_control_cheap.enable_history_data();
     solver_control_expensive.enable_history_data();
 
+    // The A block should be symmetric, unless there are free surface stabilization terms, or
+    // the user has forced the use of a different solver.
+    bool A_block_is_symmetric = true;
+    if (sim.mesh_deformation && sim.mesh_deformation->get_boundary_indicators_requiring_stabilization().empty() == false)
+      A_block_is_symmetric = false;
+    else if (sim.parameters.force_nonsymmetric_A_block_solver)
+      A_block_is_symmetric = false;
```

Comment:
It's awkward that this code block is repeated twice. Could we move that somewhere into `class Simulator`, or another central place? Or could we handle this in a different way (e.g., by having a signal that the different places that add to the matrix could attach to, and that could return a flag for "yes, we're adding nonsymmetric tems")?